### PR TITLE
draft NEP: array API standard adoption

### PR DESCRIPTION
**Pull request by rgommers**
PR-to-self for review of draft, before submitting as a PR to the main repo

Cc @shoyer, @asmeurer

### CODE REVIEW

File: numpy/version.py [hunk 1]
```diff
@@ -1,11 +0,0 @@
...
-
-__ALL__ = ['version', 'full_version', 'git_revision', 'release']
-
-vinfo = get_versions()
-version: str = vinfo["version"]
-full_version: str = vinfo['version']
-git_revision: str = vinfo['full-revisionid']
-release = 'dev0' not in version
-
-del get_versions, vinfo
```
Comment (asmeurer) [inline numpy/version.py:11]: Did you intend to delete this file?

File: doc/neps/nep-0047-array-api-standard.rst [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+The standard current does not prescribe a name for the array object itself.
+We propose to simply name it ``ndarray``. This is the most obvious name, and
+because of the separate namespace should not clash with ``numpy.ndarray``.
+
+
+Implementation
+--------------
+
+A prototype of the ``array_api`` namespace can be found in
+https://github.com/data-apis/numpy/tree/array-api/numpy/_array_api.
```
Comment (asmeurer) [inline doc/neps/nep-0047-array-api-standard.rst:432]: Perhaps this NEP should also be done on the data-apis fork?

File: doc/neps/nep-0047-array-api-standard.rst [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+An indexing expression that would return a scalar with ``ndarray``, e.g.
+``arr_2d[0, 0]``, will return a 0-D array with the new array object. There are
+several reasons for that: array scalars are largely considered a design mistake
+which no other array library copied; it works better for non-CPU libraries
+(typically arrays can live on the device, scalars live on the host); and it's
+simply a consistent design. To get a Python scalar out of a 0-D array, one can
+simply use the builtin for the type, e.g. ``float(arr_0d)``.
+
+The other `indexing modes in the standard <https://data-apis.github.io/array-api/latest/API_specification/indexing.html>`__
+do work the same as they do for ``numpy.ndarray``.
```
Comment (asmeurer) [inline doc/neps/nep-0047-array-api-standard.rst:397]: It's not clear here. Is the plan to make a separate array object that implements only the indexing that is strictly required by the standard?

File: doc/neps/nep-0047-array-api-standard.rst [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+
+The lack of advanced indexing, and boolean indexing being limited to a single
+n-D boolean array, is due to those indexing modes not being suitable for all
+types of arrays or JIT compilation. Their absence does not seem to be
```
Comment (asmeurer) [inline doc/neps/nep-0047-array-api-standard.rst:401]: Note that the spec also omits some standard indexing cases, like clipping in slices (like `a[:n]` where `n` is greater than the size of the first axis). It also doesn't specify the behavior for out of bounds integer indices.

File: doc/neps/nep-0047-array-api-standard.rst [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+  under-discussion design choices
+- Relation to existing and proposed NumPy array protocols
+  (``__array_ufunc__``, ``__array_function__``, ``__array_module__``).
+- Required improvements to existing NumPy functionality
+
+Out of scope for this NEP are:
+
+- Changes in the array API standard itself. Those are likely to come up
+  during review of this NEP, but should be upstreamed as needed and this NEP
+  subsequently updated.
```
Comment (asmeurer) [inline doc/neps/nep-0047-array-api-standard.rst:81]: This is the point that I suspect will need to be stressed the hardest.

File: numpy/version.py (hunk 1 above)
Comment (rgommers) [reply numpy/version.py:11]: argh, no - not sure what's going on there.

File: doc/neps/nep-0047-array-api-standard.rst (hunk 3 above)
Comment (rgommers) [reply doc/neps/nep-0047-array-api-standard.rst:397]: Yes - you were going to work on a Python implementation, we talked about that Monday last week?

File: doc/neps/nep-0047-array-api-standard.rst (hunk 2 above)
Comment (rgommers) [reply doc/neps/nep-0047-array-api-standard.rst:432]: I can send the PR from there, but it really doesn't matter as much I'd say. The prototype will be reusable for other libraries, the NEP is really just for NumPy.

**Comment by shoyer** [comment]
"Astropy" appears twice in the "library dependencies" figure -- was that intentional?

The specific example dependencies are also a little strange and missing some real-life connections, e.g., scikit-learn only integrates with PyTorch? I would suggest either making it precise (with all relevant dependencies) or keeping it higher level, without showing dependencies.

File: doc/neps/nep-0047-array-api-standard.rst [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+encounter any other value.
+
+
+Dtypes and casting rules
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The supported dtypes in this namespace are boolean, 8/16/32/64-bit signed and
+unsigned integer, and 32/64-bit floating-point dtypes. These will be added to
+the namespace as dtype literals with the expected names (e.g., ``bool``,
+``uint16``, ``float64``).
```
Comment (shoyer) [inline doc/neps/nep-0047-array-api-standard.rst:336]: It may be worth clarifying what the exact type of "dtype literal" objects will be for NumPy, even if the API standard leaves this as ambiguous.

For example, are these Python `type` objects? Is there any subclass/subtyping relationship between them?

File: doc/neps/nep-0047-array-api-standard.rst [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+~~~~~~~~~~~~~~~~
+
+The array object in the standard does not have methods other than dunder
+methods. The rationale for that is that not all array libraries have methods
+on their array object (e.g., TensorFlow does not). It also provides only a
+single way of doing something, rather than have functions and methods that
+are effectively duplicate.
+
+Mixing operations that may produce views (e.g., indexing, ``nonzero``,
+``where``) in combination with mutation (e.g., item or slice assignment) is
```
Comment (shoyer) [inline doc/neps/nep-0047-array-api-standard.rst:418]: `where` is probably a bad example, since I assume we only support the three argument variant `where(condition, x, y)` in the API standard.

File: doc/neps/nep-0047-array-api-standard.rst [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+            # one could special-case np.ndarray above or use np.asarray here if
+            # older numpy versions need to be supported.
+            raise ValueError("Unrecognized array input")
+
+        if len(namespaces) != 1:
+            raise ValueError(f"Multiple namespaces for array inputs: {namespaces}")
+
+        xp, = namespaces
+        if xp is None:
+            raise ValueError("The input is not  a supported array type")
```
Comment (shoyer) [inline doc/neps/nep-0047-array-api-standard.rst:506]:
```suggestion
            raise ValueError("The input is not a supported array type")
```

**Review by shoyer — APPROVED** [review]

File: doc/neps/nep-0047-array-api-standard.rst [hunk 9]
```diff
@@ -0,0 +1,534 @@
...
+Indexing
+~~~~~~~~
+
+An indexing expression that would return a scalar with ``ndarray``, e.g.
+``arr_2d[0, 0]``, will return a 0-D array with the new array object. There are
+several reasons for that: array scalars are largely considered a design mistake
+which no other array library copied; it works better for non-CPU libraries
+(typically arrays can live on the device, scalars live on the host); and it's
+simply a consistent design. To get a Python scalar out of a 0-D array, one can
+simply use the builtin for the type, e.g. ``float(arr_0d)``.
```
Comment (shoyer) [inline doc/neps/nep-0047-array-api-standard.rst:394]: A likely complaint would be that there is no way to get a Python scalar of an _unknown_ type out.

In NumPy, the idiomatic way to do this is either `arr_0d.item()` or `arr_0d[()]`, but neither is supported in the API standard.

Comment (rgommers) [reply doc/neps/nep-0047-array-api-standard.rst:394]: Good point. `[()]` wasn't thought about, should be added if all libraries support it. Most libraries have `.item()`, only tensorflow seems to use `.eval()`. A bit friendlier than empty-tuple indexing, so could add that as well.

File: doc/neps/nep-0047-array-api-standard.rst (hunk 6 above)
Comment (rgommers) [reply doc/neps/nep-0047-array-api-standard.rst:336]: I'd suggest just aliasing them to `float64` et al. in the main namespace. And having a private validation function for `dtype=` to ensure users pass in one of those aliases and not a format string for example.

Comment (rgommers) [reply doc/neps/nep-0047-array-api-standard.rst:336]: Although on reflection, I think we want this to work:
```
numpy.array_api.some_func(..., dtype=x.dtype)
```
Right now:
```
>>> x.dtype
dtype('int64')
>>> x.dtype is np.int64
False
>>> x.dtype is np.dtype(np.int64)
True
```
Making them `np.dtype` instances could make sense. Or some singleton class - that would not expose all the unnecessary dtype class attributes, but it'd be annoying wrapping.

For now I'll add: `np.array_api.float64 = np.dtype(np.float64)`. If you see a problem with that or have a better idea, please let me know.

Comment (rgommers) [reply doc/neps/nep-0047-array-api-standard.rst:336]: > Is there any subclass/subtyping relationship between them?

That's a good question. I'd prefer to keep that as simple as possible. We should figure out if we need anything from the `np.is*` set of functions. There's an open numpy issue about these dtype functions being a mess, we should probably just link to https://github.com/numpy/numpy/issues/17325 and see if we can resolve that.

Using `isinstance` is bad, something like `np.issubdtype(x.dtype, np.integer)` is worse imho.

In the standard we have `isnan` and `isinf`, maybe we should add `isinteger`, `isfloating`.

Comment (rgommers) [reply doc/neps/nep-0047-array-api-standard.rst:336]: Or maybe it's now simple enough to do:
```
def isfloating(dtype):
    return dtype in (float32, float64)

def isinteger(dtype):
    return dtype in (uint8, unit16, ..., int64)
```
in user code. maybe something worth opening an issue about on the array-api repo.

File: doc/neps/nep-0047-array-api-standard.rst (hunk 9 above)
Comment (rgommers) [reply doc/neps/nep-0047-array-api-standard.rst:394]: Actually the TF part was wrong - my TensorFlow is rusty. TF does not seem to have a good way of doing this, there's no method and indexing with an empty tuple is a no-op:
```
>>> a = tf.constant(1)
>>> type(a)
<class 'tensorflow.python.framework.ops.EagerTensor'>
>>> type(a[()])
<class 'tensorflow.python.framework.ops.EagerTensor'>
>>> b = np.array(1)
>>> type(b)
<class 'numpy.ndarray'>
>>> type(b[()])
<class 'numpy.int64'>
```

PyTorch works like TF for indexing. That all makes sense, because numpy produces array scalars, which don't exist in other libraries. And indexing shouldn't produce a Python int/float here.

PyTorch does have `.item()` which produces a Python scalar. But that's then still inconsistent with NumPy - NumPy itself arguably has no way of producing a Python scalar of unknown type.

I'm not sure if implementing `.item()` is desired or not.

Comment (shoyer) [reply doc/neps/nep-0047-array-api-standard.rst:394]: I don't think `[()]` should return a scalar. It only works that way in NumPy because array indexing returns scalars.

> But that's then still inconsistent with NumPy - NumPy itself arguably has no way of producing a Python scalar of unknown type.

Doesn't NumPy have `.item()`?

That said, maybe we don't really need `.item()` until we hear complaints about it. Unboxing is only a valid thing for some array types, and in dask.array, I believe `.item()` can actually return a `dask.delayed` object.

Comment (asmeurer) [reply doc/neps/nep-0047-array-api-standard.rst:394]: What is the purpose of even mentioning Python scalars here? It seems irrelevant to the question of indexing. Note that `arr[()]` in NumPy returns a NumPy scalar, not a Python scalar, even if the array is 0-dimensional:

```py
>>> type(np.array(1.0)[()])
<class 'numpy.float64'>
```

It might be clearer to just have a separate section that makes it clear that there are no scalar types in the array API, only 0-d arrays. That can be mentioned here in the indexing section, but I think it's about more than just indexing (indexing is just one way that you can end up with a scalar, though generally the most common). Right now this is sort of mentioned in a few places, but it's important enough to warrant a separate section.
```suggestion
simply a consistent design. To get a Python scalar out of a 0-D array, one can
simply use the builtin for the type, e.g. ``float(arr_0d)``.
```
```suggestion
simply a consistent design. To get a Python scalar out of a 0-D array, one can
simply use the builtin for the type, e.g. ``float(arr_0d)``.
```

And even then, *Python scalars* are a whole other thing. Mixing them up in the same section only confuses things, especially if we use the same terminology ("scalar") for both. The spec can probably get away with this (though IMO it should also be less ambiguous), because it doesn't have scalar types, but NumPy does, so it would help to be clear about these points and to keep them separate.

Comment (rgommers) [reply doc/neps/nep-0047-array-api-standard.rst:394]: @asmeurer good question, I'll see if a separate section makes sense. The array scalars point itself is minor, and no one will be surprised that they're excluded. The "get me a Python scalar" thing is more important, and there should be a good way of getting one.

File: doc/neps/nep-0047-array-api-standard.rst [hunk 10]
```diff
@@ -0,0 +1,539 @@
...
+unsigned integer, and 32/64-bit floating-point dtypes. These will be added to
+the namespace as dtype literals with the expected names (e.g., ``bool``,
+``uint16``, ``float64``).
+
+The most obvious omissions are the complex dtypes. The rationale for the lack
+of complex support in the first version of the array API standard is that several
+libraries (PyTorch, MXNet) are still in the process of adding support for
+complex dtypes. The next version of the standard is expected to include ``complex32``
+and ``complex64`` (see `this issue <https://github.com/data-apis/array-api/issues/102>`__
+for more details).
```
Comment (leofang) [inline doc/neps/nep-0047-array-api-standard.rst:344]: I think you meant `complex64` and `complex128`?

File: doc/neps/nep-0047-array-api-standard.rst [hunk 11]
```diff
@@ -0,0 +1,539 @@
...
+behaviour easier - it is already moving away from value-based casting for
+example.
+
+.. note::
+
+    This section needs a lot more detail, which will gradually be added when
+    the implementation progresses.
+
+
+Feedback from downstream library authors
```
Comment (leofang) [inline doc/neps/nep-0047-array-api-standard.rst:461]: My impression is for CuPy we could simply take this new array object and s/numpy/cupy 😁

File: doc/neps/nep-0047-array-api-standard.rst (hunk 9 above)
Comment (rgommers) [reply doc/neps/nep-0047-array-api-standard.rst:395]: The outcome of this discussion in the meeting was that adding empty tuple indexing would be good for consistency, even though it's a no-op.

I think there's nothing else to do here. You cannot get a Python scalar of unknown type with NumPy today, so no urgent need to invent a new way of doing that.

File: doc/neps/nep-0047-array-api-standard.rst (hunk 6 above)
Comment (rgommers) [reply doc/neps/nep-0047-array-api-standard.rst:337]: added a section on dtype objects under _Implementation_.

File: doc/neps/nep-0047-array-api-standard.rst (hunk 10 above)
Comment (rgommers) [reply doc/neps/nep-0047-array-api-standard.rst]: good catch, fixed now

File: doc/neps/nep-0047-array-api-standard.rst (hunk 11 above)
Comment (rgommers) [reply doc/neps/nep-0047-array-api-standard.rst:510]: Used that as a quote, thanks Leo:)

File: doc/neps/nep-0047-array-api-standard.rst [hunk 12]
```diff
@@ -0,0 +1,588 @@
...
+set of functions to teach to newcomers to NumPy and recommend as "best
+practice". That contrasts with NumPy's main namespace, which contains many
+functions and objects that have been superceded or we consider mistakes - but
+that we can't remove because of backwards compatibility reasons.
+
+The usage of the ``numpy.array_api`` namespace by downstream libraries is
+intended to enable them to consume multiple kinds of arrays, *without having
+to have a hard dependency on all of those array libraries*:
+
+.. image:: _static/nep-0047-library-dependencies.png
```
Comment (rgommers) [inline doc/neps/nep-0047-array-api-standard.rst:101]: I updated this figure, it's much more realistic now - shows both the current state and (in a different color) what will be made possible by the array API standard.

**Comment by rgommers** [comment]
> "Astropy" appears twice in the "library dependencies" figure -- was that intentional?
> 
> The specific example dependencies are also a little strange and missing some real-life connections, e.g., scikit-learn only integrates with PyTorch? I would suggest either making it precise (with all relevant dependencies) or keeping it higher level, without showing dependencies.

Thanks, I did this in a bit of a hurry. Fixed up.

Can you have another look @shoyer? I'd be happy to submit it in its current state.

File: doc/neps/nep-0047-array-api-standard.rst (hunk 12 above)
Comment (shoyer) [reply doc/neps/nep-0047-array-api-standard.rst:101]: Nice!

Dask looks a little sad in the current picture -- it only has a connection to NumPy. But in fact, Dask also implement a NumPy array like interface, so would also draw a few more connections:

- From Napari and Xarray to Dask. These work today.
- From Astropy/Scikit-Learn/Scikit-Image/SciPy/Einops to Dask. These could work with the new protocol (although they might make the image too cluttered!)

I'm not sure that Holoviews is a great example, since visualization is strange use-case -- if it needs to rendered via the CPU eventually, there isn't much point in using a GPU and/or distributed array backend. Maybe a library like Qutip or YT would be a better example?

Comment (rgommers) [reply doc/neps/nep-0047-array-api-standard.rst:101]: Great points, thanks. I used QuTiP, made Dask happier, and I added a "dependency on non-array library" connection type - useful for other libraries that depend on SciPy, scikit-image and Xarray. I used libraries.io to check all dependencies, I think it's accurate now.

**Comment by rgommers** [comment]
I'll merge a bunch of open PRs that add new functions to the API standard first, so https://data-apis.github.io/array-api/latest/ is as up to date as it can be, then will send a PR to NumPy for this NEP.

**Comment by rgommers** [comment]
Sent NumPy PR, so I'll close this - thanks @shoyer and @asmeurer!

**Comment by leofang** [comment]
Nice work guys!

**Comment by shoyer** [comment]
Thanks Ralf, looks great!

For reference, https://github.com/numpy/numpy/pull/18456 is the link for the NumPy PR
